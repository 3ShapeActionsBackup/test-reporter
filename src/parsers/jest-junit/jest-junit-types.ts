export interface JunitReport {
  testsuites: TestSuites
}

export interface TestSuites {
  $: {
    name: string
    tests: number
    failures: number // assertion failed
    errors: number // unhandled exception during test execution
    time: number
  }
  testsuite: TestSuite[]
}

export interface TestSuite {
  $: {
    name: string
    tests: number
    errors: number
    failures: number
    skipped: number
    time: number
    timestamp?: Date
  }
  testcase: TestCase[]
}

export interface TestCase {
  $: {
    classname: string
    file?: string
    name: string
    time: number
  }
  failure?: string[]
  skipped?: string[]
}
