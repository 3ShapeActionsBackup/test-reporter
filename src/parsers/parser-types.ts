import {TestRunResult} from '../report/test-results'

export type Annotation = {
  path: string
  start_line: number
  end_line: number
  start_column?: number
  end_column?: number
  annotation_level: 'notice' | 'warning' | 'failure'
  message: string
  title?: string
  raw_details?: string
}

export type ParseTestResult = (files: FileContent[], options: ParseOptions) => Promise<TestResult>

export type FileContent = {path: string; content: string}

export interface ParseOptions {
  annotations: boolean
  workDir: string
  trackedFiles: string[]
}

export interface TestResult {
  testRuns: TestRunResult[]
  annotations: Annotation[]
}
