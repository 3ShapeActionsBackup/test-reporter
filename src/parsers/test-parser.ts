import {Endpoints} from '@octokit/types'

type OutputParameters = Endpoints['POST /repos/:owner/:repo/check-runs']['parameters']['output']

export type ParseTestResult = (content: string) => Promise<TestResult>

export interface TestResult {
  success: boolean
  output: OutputParameters
}
