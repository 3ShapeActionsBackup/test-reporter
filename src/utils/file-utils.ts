import * as fs from 'fs'

export function getFileContent(path: string): string {
  if (!fs.existsSync(path)) {
    throw new Error(`File '${path}' not found`)
  }

  if (!fs.lstatSync(path).isFile()) {
    throw new Error(`'${path}' is not a file`)
  }

  return fs.readFileSync(path, {encoding: 'utf8'})
}

export function normalizeDirPath(path: string, trailingSeparator: boolean): string {
  if (!path) {
    return path
  }

  path = normalizeFilePath(path)
  if (trailingSeparator && !path.endsWith('/')) {
    path += '/'
  }
  return path
}

export function normalizeFilePath(path: string): string {
  if (!path) {
    return path
  }

  return path.trim().replace(/\\/g, '/')
}
