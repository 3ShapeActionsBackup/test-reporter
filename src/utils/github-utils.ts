import * as github from '@actions/github'
import {EventPayloads} from '@octokit/webhooks'

export function getCheckRunSha(): string {
  if (github.context.payload.pull_request) {
    const pr = github.context.payload.pull_request as EventPayloads.WebhookPayloadPullRequestPullRequest
    return pr.head.sha
  }

  return github.context.sha
}
