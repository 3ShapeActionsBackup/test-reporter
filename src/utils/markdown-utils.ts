export enum Align {
  Left = ':---',
  Center = ':---:',
  Right = '---:',
  None = '---'
}

export const Icon = {
  skip: '✖️', // ':heavy_multiplication_x:'
  success: '✔️', // ':heavy_check_mark:'
  fail: '❌' // ':x:'
}

export function details(summary: string, content: string): string {
  return `<details><summary>${summary}</summary>${content}</details>`
}

export function link(title: string, address: string): string {
  return `[${title}](${address})`
}

type ToString = string | number | boolean | Date
export function table(headers: ToString[], align: ToString[], ...rows: ToString[][]): string {
  const headerRow = `| ${headers.map(tableEscape).join(' | ')} |`
  const alignRow = `| ${align.join(' | ')} |`
  const contentRows = rows.map(row => `| ${row.map(tableEscape).join(' | ')} |`).join('\n')
  return [headerRow, alignRow, contentRows].join('\n')
}

export function tableEscape(content: ToString): string {
  return content.toString().replace('|', '\\|')
}
