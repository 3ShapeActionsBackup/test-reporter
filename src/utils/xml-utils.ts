const isoDateRe = /^\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z)$/

// matches dotnet duration: 00:00:00.0010000
const durationRe = /^(\d\d):(\d\d):(\d\d\.\d+)$/

export function parseAttribute(str: string | undefined): string | Date | number | undefined {
  if (str === '' || str === undefined) {
    return str
  }

  if (isoDateRe.test(str)) {
    return new Date(str)
  }

  const durationMatch = str.match(durationRe)
  if (durationMatch !== null) {
    const [_, hourStr, minStr, secStr] = durationMatch
    return (parseInt(hourStr) * 3600 + parseInt(minStr) * 60 + parseFloat(secStr)) * 1000
  }

  const num = parseFloat(str)
  if (isNaN(num)) {
    return str
  }

  return num
}
