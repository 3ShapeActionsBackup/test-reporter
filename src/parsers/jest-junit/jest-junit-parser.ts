import {Annotation, ParseOptions, TestResult} from '../test-parser'
import {parseStringPromise} from 'xml2js'

import {JunitReport, TestCase, TestSuite, TestSuites} from './jest-junit-types'
import {Align, Icon, link, table} from '../../utils/markdown-utils'
import {normalizeFilePath} from '../../utils/file-utils'
import {slug} from '../../utils/slugger'
import {parseAttribute} from '../../utils/xml-utils'

export async function parseJestJunit(content: string, options: ParseOptions): Promise<TestResult> {
  const junit = (await parseStringPromise(content, {
    attrValueProcessors: [parseAttribute]
  })) as JunitReport
  const testsuites = junit.testsuites
  const success = !(testsuites.$?.failures > 0 || testsuites.$?.errors > 0)

  return {
    success,
    output: {
      title: junit.testsuites.$.name,
      summary: getSummary(success, junit),
      annotations: options.annotations ? getAnnotations(junit, options.workDir, options.trackedFiles) : undefined
    }
  }
}

function getSummary(success: boolean, junit: JunitReport): string {
  const stats = junit.testsuites.$

  const icon = success ? Icon.success : Icon.fail
  const time = `${stats.time.toFixed(3)}s`

  const skipped = getSkippedCount(junit.testsuites)
  const failed = stats.errors + stats.failures
  const passed = stats.tests - failed - skipped

  const heading = `# ${stats.name} ${icon}`
  const headingLine = `**${stats.tests}** tests were completed in **${time}** with **${passed}** passed, **${skipped}** skipped and **${failed}** failed.`

  const suitesSummary = junit.testsuites.testsuite.map((ts, i) => {
    const skip = ts.$.skipped
    const fail = ts.$.errors + ts.$.failures
    const pass = ts.$.tests - fail - skip
    const tm = `${ts.$.time.toFixed(3)}s`
    const result = success ? Icon.success : Icon.fail
    const tsName = ts.$.name
    const tsAddr = makeSuiteSlug(i, tsName).link
    const tsNameLink = link(tsName, tsAddr)
    return [result, tsNameLink, ts.$.tests, tm, pass, fail, skip]
  })

  const summary = table(
    ['Result', 'Suite', 'Tests', 'Time', `Passed ${Icon.success}`, `Failed ${Icon.fail}`, `Skipped ${Icon.skip}`],
    [Align.Center, Align.Left, Align.Right, Align.Right, Align.Right, Align.Right, Align.Right],
    ...suitesSummary
  )

  const suites = junit.testsuites?.testsuite?.map((ts, i) => getSuiteSummary(ts, i)).join('\n')
  const suitesSection = `## Test Suites\n\n${suites}`

  return `${heading}\n${headingLine}\n${summary}\n${suitesSection}`
}

function getSkippedCount(suites: TestSuites): number {
  return suites.testsuite.reduce((sum, suite) => sum + suite.$.skipped, 0)
}

function getSuiteSummary(suite: TestSuite, index: number): string {
  const success = !(suite.$?.failures > 0 || suite.$?.errors > 0)
  const icon = success ? Icon.success : Icon.fail

  const groups: {describe: string; tests: TestCase[]}[] = []
  for (const tc of suite.testcase) {
    let grp = groups.find(g => g.describe === tc.$.classname)
    if (grp === undefined) {
      grp = {describe: tc.$.classname, tests: []}
      groups.push(grp)
    }
    grp.tests.push(tc)
  }

  const content = groups
    .map(grp => {
      const header = grp.describe !== '' ? `#### ${grp.describe}\n\n` : ''
      const tests = table(
        ['Result', 'Test', 'Time'],
        [Align.Center, Align.Left, Align.Right],
        ...grp.tests.map(tc => {
          const name = tc.$.name
          const time = `${Math.round(tc.$.time * 1000)}ms`
          const result = getTestCaseIcon(tc)
          return [result, name, time]
        })
      )

      return `${header}${tests}\n`
    })
    .join('\n')

  const tsName = suite.$.name
  const tsSlug = makeSuiteSlug(index, tsName)
  const tsNameLink = `<a id="${tsSlug.id}" href="${tsSlug.link}">${tsName}</a>`
  return `### ${tsNameLink} ${icon}\n\n${content}`
}

function getTestCaseIcon(test: TestCase): string {
  if (test.failure) return Icon.fail
  if (test.skipped) return Icon.skip
  return Icon.success
}

function makeSuiteSlug(index: number, name: string): {id: string; link: string} {
  // use "ts-$index-" as prefix to avoid slug conflicts after escaping the paths
  return slug(`ts-${index}-${name}`)
}

function getAnnotations(junit: JunitReport, workDir: string, trackedFiles: string[]): Annotation[] {
  const annotations: Annotation[] = []
  for (const suite of junit.testsuites.testsuite) {
    for (const tc of suite.testcase) {
      if (!tc.failure) {
        continue
      }
      for (const ex of tc.failure) {
        const src = exceptionThrowSource(ex, workDir, trackedFiles)
        if (src === null) {
          continue
        }
        annotations.push({
          annotation_level: 'failure',
          start_line: src.line,
          end_line: src.line,
          path: src.file,
          message: ex,
          title: `Test Failed: ${tc.$.name}`
        })
      }
    }
  }
  return annotations
}

export function exceptionThrowSource(
  ex: string,
  workDir: string,
  trackedFiles: string[]
): {file: string; line: number; column: number} | null {
  const lines = ex.split(/\r?\n/)
  const re = /\((.*):(\d+):(\d+)\)$/

  for (const str of lines) {
    const match = str.match(re)
    if (match !== null) {
      const [_, fileStr, lineStr, colStr] = match
      const filePath = normalizeFilePath(fileStr)
      const file = filePath.startsWith(workDir) ? filePath.substr(workDir.length) : filePath
      if (trackedFiles.includes(file)) {
        const line = parseInt(lineStr)
        const column = parseInt(colStr)
        return {file, line, column}
      }
    }
  }

  return null
}
