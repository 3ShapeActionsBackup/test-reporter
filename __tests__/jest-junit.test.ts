import * as fs from 'fs'
import * as path from 'path'

import {parseJestJunit} from '../src/parsers/jest-junit/jest-junit-parser'
import {ParseOptions} from '../src/parsers/parser-types'
import {normalizeFilePath} from '../src/utils/file-utils'

const fixturePath = path.join(__dirname, 'fixtures', 'jest-junit.xml')
const outputPath = path.join(__dirname, '__outputs__', 'jest-junit.md')
const xmlFixture = {
  path: normalizeFilePath(path.relative(__dirname, fixturePath)),
  content: fs.readFileSync(fixturePath, {encoding: 'utf8'})
}

describe('jest-junit tests', () => {
  it('matches report snapshot', async () => {
    const opts: ParseOptions = {
      name: 'jest tests',
      annotations: true,
      trackedFiles: ['__tests__/main.test.js', '__tests__/second.test.js', 'lib/main.js'],
      workDir: 'C:/Users/Michal/Workspace/dorny/test-check/reports/jest/'
    }

    const result = await parseJestJunit([xmlFixture], opts)
    fs.mkdirSync(path.dirname(outputPath), {recursive: true})
    fs.writeFileSync(outputPath, result?.output?.summary ?? '')

    expect(result.success).toBeFalsy()
    expect(result?.output).toMatchSnapshot()
  })
})
