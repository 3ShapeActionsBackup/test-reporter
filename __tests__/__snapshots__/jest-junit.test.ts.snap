// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jest-junit tests matches report snapshot 1`] = `
Object {
  "annotations": Array [
    Object {
      "annotation_level": "failure",
      "end_line": 10,
      "message": "Error: expect(received).toBeTruthy()

Received: false
    at Object.<anonymous> (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\__tests__\\\\main.test.js:10:21)
    at Object.asyncJestTest (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:106:37)
    at C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:45:12
    at new Promise (<anonymous>)
    at mapper (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:28:19)
    at C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:75:41
    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
      "path": "__tests__/main.test.js",
      "start_line": 10,
      "title": "Test Failed: Failing test",
    },
    Object {
      "annotation_level": "failure",
      "end_line": 2,
      "message": "Error: Some error
    at Object.throwError (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\lib\\\\main.js:2:9)
    at Object.<anonymous> (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\__tests__\\\\main.test.js:14:11)
    at Object.asyncJestTest (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:106:37)
    at C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:45:12
    at new Promise (<anonymous>)
    at mapper (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:28:19)
    at C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:75:41
    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
      "path": "lib/main.js",
      "start_line": 2,
      "title": "Test Failed: Exception in target unit",
    },
    Object {
      "annotation_level": "failure",
      "end_line": 21,
      "message": "Error: Some error
    at Object.<anonymous> (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\__tests__\\\\main.test.js:21:11)
    at Object.asyncJestTest (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:106:37)
    at C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:45:12
    at new Promise (<anonymous>)
    at mapper (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:28:19)
    at C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\queueRunner.js:75:41
    at processTicksAndRejections (internal/process/task_queues.js:97:5)",
      "path": "__tests__/main.test.js",
      "start_line": 21,
      "title": "Test Failed: Exception in test",
    },
    Object {
      "annotation_level": "failure",
      "end_line": 1,
      "message": ": Timeout - Async callback was not invoked within the 1 ms timeout specified by jest.setTimeout.Timeout - Async callback was not invoked within the 1 ms timeout specified by jest.setTimeout.Error: 
    at new Spec (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmine\\\\Spec.js:116:22)
    at new Spec (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\setup_jest_globals.js:78:9)
    at specFactory (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmine\\\\Env.js:523:24)
    at Env.it (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmine\\\\Env.js:592:24)
    at Env.it (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmineAsyncInstall.js:134:23)
    at it (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\jasmine\\\\jasmineLight.js:100:21)
    at Object.<anonymous> (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\__tests__\\\\second.test.js:1:34)
    at Runtime._execModule (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-runtime\\\\build\\\\index.js:1245:24)
    at Runtime._loadModule (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-runtime\\\\build\\\\index.js:844:12)
    at Runtime.requireModule (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-runtime\\\\build\\\\index.js:694:10)
    at jasmine2 (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-jasmine2\\\\build\\\\index.js:230:13)
    at runTestInternal (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-runner\\\\build\\\\runTest.js:380:22)
    at runTest (C:\\\\Users\\\\Michal\\\\Workspace\\\\dorny\\\\test-check\\\\reports\\\\jest\\\\node_modules\\\\jest-runner\\\\build\\\\runTest.js:472:34)",
      "path": "__tests__/second.test.js",
      "start_line": 1,
      "title": "Test Failed: Timeout test",
    },
  ],
  "summary": "# jest tests ❌
**6** tests were completed in **1.360s** with **1** passed, **1** skipped and **4** failed.
| Result | Suite | Tests | Time | Passed ✔️ | Failed ❌ | Skipped ✖️ |
| :---: | :--- | ---: | ---: | ---: | ---: | ---: |
| ❌ | [__tests__\\\\main.test.js](#ts-0-tests-main-test-js) | 4 | 0.486s | 1 | 3 | 0 |
| ❌ | [__tests__\\\\second.test.js](#ts-1-tests-second-test-js) | 2 | 0.082s | 0 | 1 | 1 |
## Test Suites

### <a id=\\"user-content-ts-0-tests-main-test-js\\" href=\\"#ts-0-tests-main-test-js\\">__tests__\\\\main.test.js</a> ❌

#### Test 1

| Result | Test | Time |
| :---: | :--- | ---: |
| ✔️ | Passing test | 1ms |

#### Test 1 › Test 1.1

| Result | Test | Time |
| :---: | :--- | ---: |
| ❌ | Failing test | 2ms |
| ❌ | Exception in target unit | 0ms |

#### Test 2

| Result | Test | Time |
| :---: | :--- | ---: |
| ❌ | Exception in test | 0ms |

### <a id=\\"user-content-ts-1-tests-second-test-js\\" href=\\"#ts-1-tests-second-test-js\\">__tests__\\\\second.test.js</a> ❌

| Result | Test | Time |
| :---: | :--- | ---: |
| ❌ | Timeout test | 4ms |
| ✖️ | Skipped test | 0ms |
",
  "title": "jest tests",
}
`;
